### PR TITLE
Supervisor: Safety Index

### PR DESCRIPTION
**Pull request by axelKingsley**
# What
Replaces the following components:
* Safety Checker
* Chain Tracker infrastructure

With "Safety Index", located in `safety` package.

# Safety Index
Safety index takes updates to Local safties for all chains, as well as Finality. It uses those signals to advance the x-safety for all chains as well.

Callers can access the Safety Index's local and cross safety indexes, and will get HeadPointers in return, which point into the database.

Safety Index manages its unsafe and safe pointers as `View`s

## Views
Views are a new wrapper around iterators which are safety-processing specific. They contain a "Local" pointer which represents the starting point of their processing, and an iterator which represents the extent of their processing.

In effect, each view holds the imperatively set local safety index, and holds the cross-safety index by using the iterator's `TraverseConditional`.

Each type of safety has its own type of conditional traversal for x-heads. Unsafe simply advances to the first non-validating executing message within the local range. Safe, on the other hand, has a more complex check against the L1 block references stored

### TraverseConditional
Iterators now support a generic way of advancing. By providing a function which acts on the iterator's state, callers can automatically advance their index as far as possible.

# Integration
* SafetyIndex is now wired in to ChainsDB
* ChainsDB triggers an unsafe-head update once done Sealing the block into the logsDB
* Debug logging emitted during `TestInteropTrivial` reveals the unsafe processor is working

## Deletion
I deleted `SafetyCheckers` and all things `HeadStorage` related, as `SafetyIndex` replaces these structures. I wrote a new `Safest` method for the `ChainsDB` to consult Safety Indexes for the API, but otherwise everything disconnected as hoped-for.

## Future Integration
We are changing the data-flow so that nodes push their data into the supervisor. When this happens, we'll also call the update safe and update finalized flows.

## Integration Results
Currently we only have unsafe data coming in because the ingestion is wrongly trying to get data from `geth`. @protolambda is working on changes for that, but in the meantime, here are some logs demonstrating that a new write to the logs database updated *one* Local Head, and triggered the Cross-Head update of two chains:
```
    safety.go:87:               DEBUG[09-26|15:42:07.955] Updated local unsafe head                role=supervisor chainID=900200 local="{LastSealedBlockHash:0xd0f6763efe80a1a7a484f17973380e2e2dc2338ecb939f4f4d83662b353aeb73 LastSealedBlockNum:33 LastSealedTimestamp:1727383328 LogsSince:0}"
    safety.go:99:               DEBUG[09-26|15:42:07.955] Updated cross unsafe head                role=supervisor chainID=900200 cross="{LastSealedBlockHash:0xd0f6763efe80a1a7a484f17973380e2e2dc2338ecb939f4f4d83662b353aeb73 LastSealedBlockNum:33 LastSealedTimestamp:1727383328 LogsSince:0}"
    safety.go:99:               DEBUG[09-26|15:42:07.955] Updated cross unsafe head                role=supervisor chainID=900201 cross="{LastSealedBlockHash:0x569d3b254cd03e00922d68b1a5899b9ff160ae7b00c87ee6f3ab51e5cc68cae0 LastSealedBlockNum:33 LastSealedTimestamp:1727383328 LogsSince:0}"
```

# Tests
* Existing/Implicit Tests. New Tests WIP.

# Indexes, Iterators, Pointers, Views, Oh My!
Here's a rundown of the different structures we are using to serve data out of the logs database:
#### The Database Itself
The Logs Database is a fixed-width append only database with regular checkpointing. The data stored is the log event data (or lack thereof) for every block in order, including detailed information about executing messages.

### Indicies
An `entrydb.EntryIdx` is a `uint64` which specifies an exact row into the database structure. With enough context, the data at that position could be read, but indexes are primitives.

### Iterators
Iterators represent a read-head on the database. They can advance forward, and can read out the various data-types stored in the database. They consume rows of the database to advance an internal state machine (called `logContext`)
**Log Contexts represent the state-relevant parameters for the iterator, including its position, last observed block-hash, etc.

Iterators become extra flexible with the use of `TraverseConditional`, which takes a function that will process the iteratorState until your given function returns an error.

### Head Pointers
`HeadPointer`s are logical references into the database. They contain the last sealed block number, and an offset called `LogsSince`. They can answer basic inequality questions about if another Head Pointer is behind it, or if it's sealed in relation to it.

### Views
`View`s are evaluators on the database, and in particular they are created to maintain "x-head" values. They do the following:
* Store "Local Head" as a `HeadPointer`
* `Process`, which advances an `Iterator` through Executing Events and runs a configured Validity Check on them, stopping when an error is encountered. The Local Head Pointer is used in these functions while the iterator processes through data, for range-checking.

Views return `Local` and `Cross` values by returning a Head Pointer, or by transforming the iterator into a Head Pointer.

### CODE REVIEW

File: op-supervisor/supervisor/backend/safety/safety.go [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+func (r *safetyIndex) advanceCrossSafe() {
+	for chainID, view := range r.safe {
+		if err := view.Process(); err != nil {
+			r.log.Error("Failed to update cross-safe view", "chain", chainID, "err", err)
+		}
+		cross, _ := r.safe[chainID].Cross()
+		r.log.Debug("Updated local safe head", "chainID", chainID, "cross", cross)
+	}
+	r.advanceFinalized()
+	// TODO prune any L2 derivedFrom entry older than the L2 finalized entry
```
Comment (protolambda) [inline op-supervisor/supervisor/backend/safety/safety.go]:
```suggestion
```
Once we're not doing the safety index fully in memory, then there's no need for pruning anymore. Having a complete historical safety-index view is actually useful. At least also to cover the span of duration for fault-proof games to determine accurate L1-input / L2-anchor context.

File: op-supervisor/supervisor/backend/safety/safety.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+			continue
+		}
+		// TODO we need to consider older cross-safe data,
+		//  if we want to finalize something at all on longer lagging finality signal.
+		// Could consider just iterating over all derivedFrom contents?
+		l1Dep := r.derivedFrom[chainID][crossSafe.LastSealedBlockHash]
+		if l1Dep.Number < r.finalizedL1.Number {
+			// TODO Temporary: truncated hashes have been replaced with full hashes
+			fullHash := common.Hash{}
+			fullHash.SetBytes(crossSafe.LastSealedBlockHash[:])
```
Comment (protolambda) [inline op-supervisor/supervisor/backend/safety/safety.go]: This can be simplified now

File: op-supervisor/supervisor/backend/safety/safety.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+		crossSafe, err := view.Cross()
+		if err != nil {
+			r.log.Info("Failed to get cross-safe data, cannot finalize", "chain", chainID, "err", err)
+			continue
+		}
+		// TODO we need to consider older cross-safe data,
+		//  if we want to finalize something at all on longer lagging finality signal.
+		// Could consider just iterating over all derivedFrom contents?
+		l1Dep := r.derivedFrom[chainID][crossSafe.LastSealedBlockHash]
+		if l1Dep.Number < r.finalizedL1.Number {
```
Comment (protolambda) [inline op-supervisor/supervisor/backend/safety/safety.go]: Would be nice to ticket this

File: op-supervisor/supervisor/backend/safety/views.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+func (vi *View) UpdateLocal(at eth.L1BlockRef, ref eth.L2BlockRef) error {
+	vi.localView = heads.HeadPointer{
+		LastSealedBlockHash: ref.Hash,
+		LastSealedBlockNum:  ref.Number,
+		//LastSealedTimestamp: ref.Time,
+		LogsSince: 0,
+	}
+	vi.localDerivedFrom = at
+
+	// TODO(devnet 2) reorg check against existing DB
```
Comment (protolambda) [inline op-supervisor/supervisor/backend/safety/views.go]: There's a bunch of TODOs here related to reorg checks, which we should reference the github issue in.

**Review by protolambda — APPROVED** [review]
Good progress, will need more testing next week, but already closer to getting e2e safety-increments running correctly.

File: op-supervisor/supervisor/backend/safety/safety.go (hunk 3 above)
Comment (axelKingsley) [reply op-supervisor/supervisor/backend/safety/safety.go]: https://github.com/ethereum-optimism/optimism/issues/12184

**Pull request merged**